### ISSUE TITLE
Add a blacklist inverse of DeviceAllow

### ISSUE THREAD

**Issue by Melab** [systemd/systemd#31695]
### Is your feature request related to a problem? Please describe

The `DeviceAllow` resource control setting is currently available for whitelisting the devices that applied-to processes may accesses. A setting, perhaps called `DeviceDeny`, should be added to accomodate the case where applied-to processes should have access to all devices except those specified.

**Comment by Winterhuman** [comment]
I've opened an issue before for supporting this: https://github.com/systemd/systemd/issues/30780

However, devices also take options afterwards, and the consensus is that it would be very difficult to implement in practice and repeat what cgroups v1 did.

**Comment by Melab** [comment]
> I've opened an issue before for supporting this: #30780
> 
> However, devices also take options afterwards, and the consensus is that it would be very difficult to implement in practice and repeat what cgroups v1 did.

I'm sorry, but I don't understand:

1. Do devices *not* take options with `devices.allow`/`DeviceAllow`?
2. Of what relevance is the difference between cgroups version 1 and version 2? Why would you need to "repeat what cgroups v1 did"? I can't tell what that would mean. The kernel does what cgroups does. systemd wouldn't be doing what cgroups does.

**Comment by Winterhuman** [comment]
Specifically this part: "cgroup v1 had both devices.allow and devices.deny, that was difficult to evaluate. (Re)adding a denylist sounds like going in circles to me."

What I meant was that adding `DeviceDeny` would mean you'd need to negate devices, but also, you'd need to negate device options on top of that, which further complicates the implementation (that's how I understood the conversation in that issue).

**Comment by Melab** [comment]
> Specifically this part: "cgroup v1 had both devices.allow and devices.deny, that was difficult to evaluate. (Re)adding a denylist sounds like going in circles to me."

Does cgroups v2 does not have the ability to deny read/write access to devices? Isn't the kernel the one evaluating the cgroups data in `/sys/fs/cgroup`, not systemd?

> What I meant was that adding `DeviceDeny` would mean you'd need to negate devices, but also, you'd need to negate device options on top of that, which further complicates the implementation (that's how I understood the conversation in that issue at least).

What are these "options" that you are talking about?

**Comment by Winterhuman** [comment]
e.g. `DeviceAllow=char-pts r` would give read-only access to PTS devices, while `DeviceDeny=char-pts r` would only allow creating and writing to PTS devices, so if both `DeviceAllow=` and `DeviceDeny=` are used you'd need to evaluate both for any conflicting options, possibly doing so based on the order they're specified in.

As for the cgroups v2 question, I honestly have no clue; I'm just going based on the comments from the developers

**Comment by Melab** [comment]
> e.g. `DeviceAllow=char-pts r` would give read-only access to PTS devices, while `DeviceDeny=char-pts r` would only allow creating and writing to PTS devices, so if both `DeviceAllow=` and `DeviceDeny=` are used you'd need to evaluate both for any conflicting options, possibly doing so based on the order they're specified in.

That sounds relatively easy. AppArmor, as an example, calculates the allowed set of actions by subtracting the blacklisted actions from the whitelisted actions. So, if an action is in both lists, it is blocked by AppArmor.

> As for the cgroups v2 question, I honestly have no clue; I'm just going based on the comments from the developers

I researched it and discovered that eBPF programs are what are used by the kernel to filter device access in cgroups version 2. That sound like a considerable **downgrade**.

**Comment by poettering** [comment]
Duplicate of #19222
